src/backend/catalog/storage_pending_deletes_redo.c

```diff
@@ -0,0 +1,344 @@
+		 * case we can't rely on xid status of that frozen transactions.
+		 * Second point is that there is no way that clog would be truncated
+		 * when transaction is in progress, so it's either been committed or
+		 * aborted before that.
+		 */
+
+		if (TransactionIdPrecedes(pd->xid, oldest_xid))
+			elog(WARNING,
+				 "Prevented adding node for XLOG_PENDING_DELETE "
+				 "record for xid: %u, oldestXid: %u",
+				 pd->xid, oldest_xid);
+		else
+		{
+			XLogRecPtr	result;
+			XidStatus	status = TransactionIdGetStatus(pd->xid, &result);
+
+			if (status == TRANSACTION_STATUS_IN_PROGRESS)
+				PdlRedoAdd(pd);
+			else
+				elog(WARNING,
+					 "Prevented adding node for XLOG_PENDING_DELETE "
+					 "record for xid: %u, status: %d",
+					 pd->xid, status);
+		}
+	}
+}
+
+static void
+PdlRedoRemove(TransactionId xid)
+{
+	if ((xid == InvalidTransactionId) ||
+		(NULL == pendingDeletesRedo))
+		return;
+
+	PendingDeleteHtabNode *entry = (PendingDeleteHtabNode *)
+		hash_search(pendingDeletesRedo, &xid, HASH_REMOVE, NULL);
+
+	if (entry)
+		list_free_deep(entry->relnode_list);
+}
+
+/*
+ * This function removes pending delete nodes from redo hash-table
+ * (pendingDeleteRedo) for a given transaction identified by it's xid and
+ * sub-transactions (if there are).
+ */
+void
+PdlRedoRemoveTree(TransactionId xid,
+				  TransactionId *sub_xids, int nsubxacts)
+{
+	if (PdlTrackingDisabled())
+		return;
+
+	for (int i = 0; i < nsubxacts; i++)
+		PdlRedoRemove(sub_xids[i]);
+
+	PdlRedoRemove(xid);
+}
+
+/*
+ * This function serializes the contents of hash table entry into a structure
+ * suitable to pass into DropRelationFiles() functions.
+ */
+static RelFileNodePendingDelete *
+PdlRedoPrepareArrayForDrop(PendingDeleteHtabNode *hnode, int *ndelrels)
+{
+	ListCell   *cell;
+
+	foreach(cell, hnode->relnode_list)
+	{
+		RelFileNodePendingDelete *pending_delete_node =
+			(RelFileNodePendingDelete *) lfirst(cell);
+		ListCell   *i_cell = lnext(cell);
+		ListCell   *i_cell_prev = cell;
+
+		while (i_cell)
+		{
+			ListCell   *i_cell_next = lnext(i_cell);
+			RelFileNodePendingDelete *i_relnode =
+				(RelFileNodePendingDelete *) lfirst(i_cell);
+
+			if (RelFileNodeEquals(pending_delete_node->node, i_relnode->node))
+			{
+				elog(DEBUG1,
+					 "Duplicate pending delete node found: "
+					 "(rel: (%u: %u: %u); xid: %u)",
+					 pending_delete_node->node.spcNode,
+					 pending_delete_node->node.dbNode,
+					 pending_delete_node->node.relNode,
+					 hnode->xid);
+
+				hnode->relnode_list =
+					list_delete_cell(hnode->relnode_list, i_cell, i_cell_prev);
+				pfree(i_relnode);
+			}
+			else
+				i_cell_prev = i_cell;
+
+			i_cell = i_cell_next;
+		}
+	}
+
+	*ndelrels = list_length(hnode->relnode_list);
+
+	if (*ndelrels <= 0)
+	{
+		elog(WARNING, "Empty list for xid: %u", hnode->xid);
+		return NULL;
+	}
+
+	RelFileNodePendingDelete *delrels = (RelFileNodePendingDelete *)
+		palloc((*ndelrels) * sizeof(RelFileNodePendingDelete));
```

Comment:
Let's use `sizeof(*delrels)` as in other places to be sure that the pointer type matches the type in sizeof
```suggestion
		palloc((*ndelrels) * sizeof(*delrels));
```